### ISSUE TITLE
Library is not working with large tables

### ISSUE THREAD

**Issue by jmanteau** [community-fabric/python-ipfabric#178]
Today _ipf_pager is "ready" to support pagination but both query and fetch_all doesn't implement the logic.

It means that we get this kind of behavior on large tables :

```
ipdb> ipf.technology.addressing.mac_table.count()
6498944
ipdb> mac_table= ipf.technology.addressing.mac_table.all()
*** httpx.ReadTimeout: The read operation timed out
```

@sdargoeuves : FYI

**Comment by justinjeffery-ipf** [comment]
Please provide what version of the SDK and IP Fabric you are using.  

`httpx.ReadTimeout` indicates a timeout from a single response from the server, not the entire time it takes to get all 6,498,944/1,000 requests.

"The read timeout specifies the maximum duration to wait for a chunk of data to be received (for example, a chunk of the response body). If HTTPX is unable to receive data within this time frame, a ReadTimeout exception is raised."

Please try: 
`ipf = IPFClient(timeout=15)`

This has been tested and should be implemented correctly:
```
>>> rcount = ipf.get_count('tables/networks/routes')
>>> rcount
164074
>>> routes = ipf.fetch_all('tables/networks/routes')
>>> len(routes)
164074
```

**Comment by sdargoeuves** [comment]
@jmanteau Could you please let us know if you are still facing this httpx.ReadTimeout issue?

**Comment by jmanteau** [comment]
Hello,

Sorry, I was on sick leave.

I have re-done the test with python 3.10.3 and latest SDK version (6.0.5).
Including the timeout in the client initialization does not work for me:

```
In [10]: ipf = IPFClient(IPF_ADDR, username=IPF_USERNAME, password=IPF_PASSWORD,timeout=15)
Specified API or SDK Version (6.0.5) is greater then OS API Version. Using OS Version:  (5.0)
Retrieving only loaded snapshots. To load all snapshots set `unloaded` to True.

In [11]: print(ipf.technology.addressing.mac_table.count())
5640434

In [12]: mac_table= ipf.technology.addressing.mac_table.all()


---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_exceptions.py:8, in map_exceptions(map)
      7 try:
----> 8     yield
      9 except Exception as exc:  # noqa: PIE786

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/backends/sync.py:26, in SyncStream.read(self, max_bytes, timeout)
     25 self._sock.settimeout(timeout)
---> 26 return self._sock.recv(max_bytes)

File ~/.pyenv/versions/3.10.3/lib/python3.10/ssl.py:1258, in SSLSocket.recv(self, buflen, flags)
   1255         raise ValueError(
   1256             "non-zero flags not allowed in calls to recv() on %s" %
   1257             self.__class__)
-> 1258     return self.read(buflen)
   1259 else:

File ~/.pyenv/versions/3.10.3/lib/python3.10/ssl.py:1131, in SSLSocket.read(self, len, buffer)
   1130     else:
-> 1131         return self._sslobj.read(len)
   1132 except SSLError as x:

TimeoutError: The read operation timed out

During handling of the above exception, another exception occurred:

ReadTimeout                               Traceback (most recent call last)
File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_transports/default.py:60, in map_httpcore_exceptions()
     59 try:
---> 60     yield
     61 except Exception as exc:  # noqa: PIE-786

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_transports/default.py:218, in HTTPTransport.handle_request(self, request)
    217 with map_httpcore_exceptions():
--> 218     resp = self._pool.handle_request(req)
    220 assert isinstance(resp.stream, typing.Iterable)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:253, in ConnectionPool.handle_request(self, request)
    252     self.response_closed(status)
--> 253     raise exc
    254 else:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:237, in ConnectionPool.handle_request(self, request)
    236 try:
--> 237     response = connection.handle_request(request)
    238 except ConnectionNotAvailable:
    239     # The ConnectionNotAvailable exception is a special case, that
    240     # indicates we need to retry the request on a new connection.
   (...)
    244     # might end up as an HTTP/2 connection, but which actually ends
    245     # up as HTTP/1.1.

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http_proxy.py:313, in TunnelHTTPConnection.handle_request(self, request)
    312         self._connected = True
--> 313 return self._connection.handle_request(request)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:105, in HTTP11Connection.handle_request(self, request)
    104     self._response_closed()
--> 105 raise exc

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:84, in HTTP11Connection.handle_request(self, request)
     76 with Trace(
     77     "http11.receive_response_headers", request, kwargs
     78 ) as trace:
     79     (
     80         http_version,
     81         status,
     82         reason_phrase,
     83         headers,
---> 84     ) = self._receive_response_headers(**kwargs)
     85     trace.return_value = (
     86         http_version,
     87         status,
     88         reason_phrase,
     89         headers,
     90     )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:148, in HTTP11Connection._receive_response_headers(self, request)
    147 while True:
--> 148     event = self._receive_event(timeout=timeout)
    149     if isinstance(event, h11.Response):

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:177, in HTTP11Connection._receive_event(self, timeout)
    176 if event is h11.NEED_DATA:
--> 177     data = self._network_stream.read(
    178         self.READ_NUM_BYTES, timeout=timeout
    179     )
    181     # If we feed this case through h11 we'll raise an exception like:
    182     #
    183     #     httpcore.RemoteProtocolError: can't handle event type
   (...)
    187     # perspective. Instead we handle this case distinctly and treat
    188     # it as a ConnectError.

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/backends/sync.py:24, in SyncStream.read(self, max_bytes, timeout)
     23 exc_map = {socket.timeout: ReadTimeout, socket.error: ReadError}
---> 24 with map_exceptions(exc_map):
     25     self._sock.settimeout(timeout)

File ~/.pyenv/versions/3.10.3/lib/python3.10/contextlib.py:153, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
    152 try:
--> 153     self.gen.throw(typ, value, traceback)
    154 except StopIteration as exc:
    155     # Suppress StopIteration *unless* it's the same exception that
    156     # was passed to throw().  This prevents a StopIteration
    157     # raised inside the "with" statement from being suppressed.

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_exceptions.py:12, in map_exceptions(map)
     11     if isinstance(exc, from_exc):
---> 12         raise to_exc(exc)
     13 raise

ReadTimeout: The read operation timed out

The above exception was the direct cause of the following exception:

ReadTimeout                               Traceback (most recent call last)
Cell In [12], line 1
----> 1 mac_table= ipf.technology.addressing.mac_table.all()

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/models.py:76, in Table.all(self, columns, filters, attr_filters, snapshot_id, reports, sort)
     57 def all(
     58     self,
     59     columns: list = None,
   (...)
     64     sort: Optional[dict] = None,
     65 ):
     66     """
     67     Gets all data from corresponding endpoint
     68     :param columns: list: Optional columns to return, default is all
   (...)
     74     :return: list: List of Dictionaries
     75     """
---> 76     return self.client.fetch_all(
     77         self.endpoint,
     78         columns=columns,
     79         filters=filters,
     80         attr_filters=attr_filters,
     81         snapshot_id=snapshot_id,
     82         reports=reports,
     83         sort=sort,
     84         snapshot=self.snapshot,
     85     )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/client.py:28, in check_format.<locals>.wrapper(self, url, *args, **kwargs)
     26 url = path[r.end() :] if r else path
     27 url = url[1:] if url[0] == "/" else url
---> 28 return func(self, url, *args, **kwargs)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/client.py:135, in IPFClient.fetch_all(self, url, columns, filters, snapshot_id, reports, sort, attr_filters, snapshot)
    133 payload = dict(columns=columns or self._get_columns(url), snapshot=snapshot_id or self.snapshot_id)
    134 payload = self._check_payload(payload, snapshot, filters, reports, sort, attr_filters)
--> 135 return self._ipf_pager(url, payload)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

    [... skipping similar frames: IPFabricAPI._ipf_pager at line 309 (570 times)]

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:304, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    301 data = data or list()
    303 payload["pagination"] = dict(limit=limit, start=start)
--> 304 r = self.post(url, json=payload)
    305 r.raise_for_status()
    306 r_data = r.json()["data"]

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:1130, in Client.post(self, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)
   1109 def post(
   1110     self,
   1111     url: URLTypes,
   (...)
   1123     extensions: typing.Optional[dict] = None,
   1124 ) -> Response:
   1125     """
   1126     Send a `POST` request.
   1127 
   1128     **Parameters**: See `httpx.request`.
   1129     """
-> 1130     return self.request(
   1131         "POST",
   1132         url,
   1133         content=content,
   1134         data=data,
   1135         files=files,
   1136         json=json,
   1137         params=params,
   1138         headers=headers,
   1139         cookies=cookies,
   1140         auth=auth,
   1141         follow_redirects=follow_redirects,
   1142         timeout=timeout,
   1143         extensions=extensions,
   1144     )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:815, in Client.request(self, method, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)
    800     warnings.warn(message, DeprecationWarning)
    802 request = self.build_request(
    803     method=method,
    804     url=url,
   (...)
    813     extensions=extensions,
    814 )
--> 815 return self.send(request, auth=auth, follow_redirects=follow_redirects)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:902, in Client.send(self, request, stream, auth, follow_redirects)
    894 follow_redirects = (
    895     self.follow_redirects
    896     if isinstance(follow_redirects, UseClientDefault)
    897     else follow_redirects
    898 )
    900 auth = self._build_request_auth(request, auth)
--> 902 response = self._send_handling_auth(
    903     request,
    904     auth=auth,
    905     follow_redirects=follow_redirects,
    906     history=[],
    907 )
    908 try:
    909     if not stream:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:930, in Client._send_handling_auth(self, request, auth, follow_redirects, history)
    927 request = next(auth_flow)
    929 while True:
--> 930     response = self._send_handling_redirects(
    931         request,
    932         follow_redirects=follow_redirects,
    933         history=history,
    934     )
    935     try:
    936         try:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:967, in Client._send_handling_redirects(self, request, follow_redirects, history)
    964 for hook in self._event_hooks["request"]:
    965     hook(request)
--> 967 response = self._send_single_request(request)
    968 try:
    969     for hook in self._event_hooks["response"]:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:1003, in Client._send_single_request(self, request)
    998     raise RuntimeError(
    999         "Attempted to send an async request with a sync Client instance."
   1000     )
   1002 with request_context(request=request):
-> 1003     response = transport.handle_request(request)
   1005 assert isinstance(response.stream, SyncByteStream)
   1007 response.request = request

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_transports/default.py:217, in HTTPTransport.handle_request(self, request)
    203 assert isinstance(request.stream, SyncByteStream)
    205 req = httpcore.Request(
    206     method=request.method,
    207     url=httpcore.URL(
   (...)
    215     extensions=request.extensions,
    216 )
--> 217 with map_httpcore_exceptions():
    218     resp = self._pool.handle_request(req)
    220 assert isinstance(resp.stream, typing.Iterable)

File ~/.pyenv/versions/3.10.3/lib/python3.10/contextlib.py:153, in _GeneratorContextManager.__exit__(self, typ, value, traceback)
    151     value = typ()
    152 try:
--> 153     self.gen.throw(typ, value, traceback)
    154 except StopIteration as exc:
    155     # Suppress StopIteration *unless* it's the same exception that
    156     # was passed to throw().  This prevents a StopIteration
    157     # raised inside the "with" statement from being suppressed.
    158     return exc is not value

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_transports/default.py:77, in map_httpcore_exceptions()
     74     raise
     76 message = str(exc)
---> 77 raise mapped_exc(message) from exc

ReadTimeout: The read operation timed out```

**Comment by jmanteau** [comment]
Hello,

For the first line, I have this output:
```
   ...: ipf = IPFClient(IPF_ADDR, username=IPF_USERNAME, password=IPF_PASSWORD,snapshot_id='$prev', timeout=None)

Specified API or SDK Version (6.0.5) is greater then OS API Version. Using OS Version:  (5.0)
Retrieving only loaded snapshots. To load all snapshots set `unloaded` to True.


In [3]: print(ipf.technology.addressing.mac_table.count())
   ...:
5836645

In [4]: time mac_table= ipf.technology.addressing.mac_table.all()
   ...:
---------------------------------------------------------------------------
RecursionError                            Traceback (most recent call last)
File <timed exec>:1

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/models.py:76, in Table.all(self, columns, filters, attr_filters, snapshot_id, reports, sort)
     57 def all(
     58     self,
     59     columns: list = None,
   (...)
     64     sort: Optional[dict] = None,
     65 ):
     66     """
     67     Gets all data from corresponding endpoint
     68     :param columns: list: Optional columns to return, default is all
   (...)
     74     :return: list: List of Dictionaries
     75     """
---> 76     return self.client.fetch_all(
     77         self.endpoint,
     78         columns=columns,
     79         filters=filters,
     80         attr_filters=attr_filters,
     81         snapshot_id=snapshot_id,
     82         reports=reports,
     83         sort=sort,
     84         snapshot=self.snapshot,
     85     )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/client.py:28, in check_format.<locals>.wrapper(self, url, *args, **kwargs)
     26 url = path[r.end() :] if r else path
     27 url = url[1:] if url[0] == "/" else url
---> 28 return func(self, url, *args, **kwargs)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/client.py:135, in IPFClient.fetch_all(self, url, columns, filters, snapshot_id, reports, sort, attr_filters, snapshot)
    133 payload = dict(columns=columns or self._get_columns(url), snapshot=snapshot_id or self.snapshot_id)
    134 payload = self._check_payload(payload, snapshot, filters, reports, sort, attr_filters)
--> 135 return self._ipf_pager(url, payload)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

    [... skipping similar frames: IPFabricAPI._ipf_pager at line 309 (2951 times)]

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:304, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    301 data = data or list()
    303 payload["pagination"] = dict(limit=limit, start=start)
--> 304 r = self.post(url, json=payload)
    305 r.raise_for_status()
    306 r_data = r.json()["data"]

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:1130, in Client.post(self, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)
   1109 def post(
   1110     self,
   1111     url: URLTypes,
   (...)
   1123     extensions: typing.Optional[dict] = None,
   1124 ) -> Response:
   1125     """
   1126     Send a `POST` request.
   1127
   1128     **Parameters**: See `httpx.request`.
   1129     """
-> 1130     return self.request(
   1131         "POST",
   1132         url,
   1133         content=content,
   1134         data=data,
   1135         files=files,
   1136         json=json,
   1137         params=params,
   1138         headers=headers,
   1139         cookies=cookies,
   1140         auth=auth,
   1141         follow_redirects=follow_redirects,
   1142         timeout=timeout,
   1143         extensions=extensions,
   1144     )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:815, in Client.request(self, method, url, content, data, files, json, params, headers, cookies, auth, follow_redirects, timeout, extensions)
    800     warnings.warn(message, DeprecationWarning)
    802 request = self.build_request(
    803     method=method,
    804     url=url,
   (...)
    813     extensions=extensions,
    814 )
--> 815 return self.send(request, auth=auth, follow_redirects=follow_redirects)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:916, in Client.send(self, request, stream, auth, follow_redirects)
    914 except BaseException as exc:
    915     response.close()
--> 916     raise exc

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:910, in Client.send(self, request, stream, auth, follow_redirects)
    908 try:
    909     if not stream:
--> 910         response.read()
    912     return response
    914 except BaseException as exc:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_models.py:792, in Response.read(self)
    788 """
    789 Read and return the response content.
    790 """
    791 if not hasattr(self, "_content"):
--> 792     self._content = b"".join(self.iter_bytes())
    793 return self._content

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_models.py:810, in Response.iter_bytes(self, chunk_size)
    808 chunker = ByteChunker(chunk_size=chunk_size)
    809 with request_context(request=self._request):
--> 810     for raw_bytes in self.iter_raw():
    811         decoded = decoder.decode(raw_bytes)
    812         for chunk in chunker.decode(decoded):

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_models.py:868, in Response.iter_raw(self, chunk_size)
    865 chunker = ByteChunker(chunk_size=chunk_size)
    867 with request_context(request=self._request):
--> 868     for raw_stream_bytes in self.stream:
    869         self._num_bytes_downloaded += len(raw_stream_bytes)
    870         for chunk in chunker.decode(raw_stream_bytes):

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_client.py:123, in BoundSyncStream.__iter__(self)
    122 def __iter__(self) -> typing.Iterator[bytes]:
--> 123     for chunk in self._stream:
    124         yield chunk

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_transports/default.py:104, in ResponseStream.__iter__(self)
    102 def __iter__(self) -> typing.Iterator[bytes]:
    103     with map_httpcore_exceptions():
--> 104         for part in self._httpcore_stream:
    105             yield part

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/connection_pool.py:346, in ConnectionPoolByteStream.__iter__(self)
    345 def __iter__(self) -> Iterator[bytes]:
--> 346     for part in self._stream:
    347         yield part

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:300, in HTTP11ConnectionByteStream.__iter__(self)
    295 except BaseException as exc:
    296     # If we get an exception while streaming the response,
    297     # we want to close the response (and possibly the connection)
    298     # before raising that exception.
    299     self.close()
--> 300     raise exc

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:293, in HTTP11ConnectionByteStream.__iter__(self)
    291 try:
    292     with Trace("http11.receive_response_body", self._request, kwargs):
--> 293         for chunk in self._connection._receive_response_body(**kwargs):
    294             yield chunk
    295 except BaseException as exc:
    296     # If we get an exception while streaming the response,
    297     # we want to close the response (and possibly the connection)
    298     # before raising that exception.

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:165, in HTTP11Connection._receive_response_body(self, request)
    162 timeout = timeouts.get("read", None)
    164 while True:
--> 165     event = self._receive_event(timeout=timeout)
    166     if isinstance(event, h11.Data):
    167         yield bytes(event.data)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpcore/_sync/http11.py:174, in HTTP11Connection._receive_event(self, timeout)
    172 while True:
    173     with map_exceptions({h11.RemoteProtocolError: RemoteProtocolError}):
--> 174         event = self._h11_state.next_event()
    176     if event is h11.NEED_DATA:
    177         data = self._network_stream.read(
    178             self.READ_NUM_BYTES, timeout=timeout
    179         )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/h11/_connection.py:425, in Connection.next_event(self)
    423     raise RemoteProtocolError("Can't receive data when peer state is ERROR")
    424 try:
--> 425     event = self._extract_next_receive_event()
    426     if event not in [NEED_DATA, PAUSED]:
    427         self._process_event(self.their_role, event)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/h11/_connection.py:367, in Connection._extract_next_receive_event(self)
    365     return PAUSED
    366 assert self._reader is not None
--> 367 event = self._reader(self._receive_buffer)
    368 if event is None:
    369     if not self._receive_buffer and self._receive_buffer_closed:
    370         # In some unusual cases (basically just HTTP/1.0 bodies), EOF
    371         # triggers an actual protocol event; in that case, we want to
    372         # return that event, and then the state will change and we'll
    373         # get called again to generate the actual ConnectionClosed().

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/h11/_readers.py:167, in ChunkedReader.__call__(self, buf)
    165     if self._bytes_in_chunk == 0:
    166         self._reading_trailer = True
--> 167         return self(buf)
    168     chunk_start = True
    169 else:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/h11/_readers.py:142, in ChunkedReader.__call__(self, buf)
    140     if lines is None:
    141         return None
--> 142     return EndOfMessage(headers=list(_decode_header_lines(lines)))
    143 if self._bytes_to_discard > 0:
    144     data = buf.maybe_extract_at_most(self._bytes_to_discard)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/h11/_events.py:55, in _EventBundle.__init__(self, **kwargs)
     52 # Special handling for some fields
     54 if "headers" in self.__dict__:
---> 55     self.headers = _headers.normalize_and_validate(
     56         self.headers, _parsed=_parsed
     57     )
     59 if not _parsed:
     60     for field in ["method", "target", "http_version", "reason"]:

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/h11/_headers.py:166, in normalize_and_validate(headers, _parsed)
    164     else:
    165         new_headers.append((raw_name, name, value))
--> 166 return Headers(new_headers)

RecursionError: maximum recursion depth exceeded
```

I'll try the second line today (the issue is I shall not disconnect for at least two hours)

**Comment by jmanteau** [comment]
```
In [7]: ipf = IPFClient(IPF_ADDR, username=IPF_USERNAME, password=IPF_PASSWORD, snapshot_id='$prev',  timeout=Timeout(True, read=None))
Python-dotenv could not parse statement starting at line 11
Python-dotenv could not parse statement starting at line 11
Specified API or SDK Version (6.0.5) is greater then OS API Version. Using OS Version:  (5.0)
Retrieving only loaded snapshots. To load all snapshots set `unloaded` to True.

In [8]: print(ipf.technology.addressing.mac_table.count())
   ...:
6457892

In [9]: time mac_table= ipf.technology.addressing.mac_table.all()
   ...:
---------------------------------------------------------------------------
HTTPStatusError                           Traceback (most recent call last)
File <timed exec>:1

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/models.py:76, in Table.all(self, columns, filters, attr_filters, snapshot_id, reports, sort)
     57 def all(
     58     self,
     59     columns: list = None,
   (...)
     64     sort: Optional[dict] = None,
     65 ):
     66     """
     67     Gets all data from corresponding endpoint
     68     :param columns: list: Optional columns to return, default is all
   (...)
     74     :return: list: List of Dictionaries
     75     """
---> 76     return self.client.fetch_all(
     77         self.endpoint,
     78         columns=columns,
     79         filters=filters,
     80         attr_filters=attr_filters,
     81         snapshot_id=snapshot_id,
     82         reports=reports,
     83         sort=sort,
     84         snapshot=self.snapshot,
     85     )

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/client.py:28, in check_format.<locals>.wrapper(self, url, *args, **kwargs)
     26 url = path[r.end() :] if r else path
     27 url = url[1:] if url[0] == "/" else url
---> 28 return func(self, url, *args, **kwargs)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/client.py:135, in IPFClient.fetch_all(self, url, columns, filters, snapshot_id, reports, sort, attr_filters, snapshot)
    133 payload = dict(columns=columns or self._get_columns(url), snapshot=snapshot_id or self.snapshot_id)
    134 payload = self._check_payload(payload, snapshot, filters, reports, sort, attr_filters)
--> 135 return self._ipf_pager(url, payload)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

    [... skipping similar frames: IPFabricAPI._ipf_pager at line 309 (2645 times)]

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:309, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    307 data.extend(r_data)
    308 if limit == len(r_data):
--> 309     self._ipf_pager(url, payload, data, limit=limit, start=start + limit)
    310 return data

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/ipfabric/api.py:305, in IPFabricAPI._ipf_pager(self, url, payload, data, limit, start)
    303 payload["pagination"] = dict(limit=limit, start=start)
    304 r = self.post(url, json=payload)
--> 305 r.raise_for_status()
    306 r_data = r.json()["data"]
    307 data.extend(r_data)

File ~/.pyenv/versions/3.10.3/lib/python3.10/site-packages/httpx/_models.py:736, in Response.raise_for_status(self)
    734 error_type = error_types.get(status_class, "Invalid status code")
    735 message = message.format(self, error_type=error_type)
--> 736 raise HTTPStatusError(message, request=request, response=self)

HTTPStatusError: Server error '504 Gateway Time-out' for url 'https://smm88-dkr-ipfabric-1.mgt.airbus.corp/api/v5.0/tables/addressing/mac'
For more information check: https://httpstatuses.com/504

```

**Comment by justinjeffery-ipf** [comment]
Thank you.  The RecursionError helped narrow it down.  Released [v6.0.6](https://github.com/community-fabric/python-ipfabric/releases/tag/v6.0.6) with a fix.  You may still need the `timeout=None` in some testing I did.

Please note that v6.X Python SDK has not been tested or is guaranteed to be backwards compatible with IP Fabric v5.X.  We are expecting v6.0.1 to be released very soon.
